quil-rs/src/instruction/gate.rs

```diff
@@ -0,0 +1,222 @@
+                actual: alt_params.len(),
+            });
+        }
+        self.modifiers.insert(0, GateModifier::Forked);
+        self.qubits.insert(0, fork_qubit);
+        self.parameters.extend(alt_params);
+        Ok(self)
+    }
+}
+
+impl fmt::Display for Gate {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let parameter_str = get_expression_parameter_string(&self.parameters);
+
+        let qubit_str = format_qubits(&self.qubits);
+        let modifier_str = &self
+            .modifiers
+            .iter()
+            .map(|m| format!("{m} "))
+            .collect::<Vec<String>>()
+            .join("");
+        write!(
+            f,
+            "{}{}{} {}",
+            modifier_str, self.name, parameter_str, qubit_str
+        )
+    }
+}
+
+impl fmt::Display for GateModifier {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use GateModifier::*;
+        write!(
+            f,
+            "{}",
+            match self {
+                Controlled => "CONTROLLED",
+                Dagger => "DAGGER",
+                Forked => "FORKED",
+            }
+        )
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum GateSpecification {
+    Matrix(Vec<Vec<Expression>>),
+    Permutation(Vec<u64>),
+}
+
+impl fmt::Display for GateSpecification {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            GateSpecification::Matrix(matrix) => {
+                for row in matrix {
+                    writeln!(
+                        f,
+                        "\t{}",
+                        row.iter()
+                            .map(|cell| format!("{cell}"))
+                            .collect::<Vec<String>>()
+                            .join(", ")
+                    )?;
+                }
+            }
+            GateSpecification::Permutation(permutation) => {
+                writeln!(
+                    f,
+                    "\t{}",
+                    permutation
+                        .iter()
+                        .map(|i| format!("{i}"))
+                        .collect::<Vec<String>>()
+                        .join(", ")
+                )?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct GateDefinition {
+    pub name: String,
+    pub parameters: Vec<String>,
+    pub specification: GateSpecification,
+}
+
+impl fmt::Display for GateDefinition {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let parameter_str = match self.parameters.is_empty() {
+            true => String::new(),
+            false => format!(
+                "({})",
+                self.parameters
+                    .iter()
+                    .map(|p| p.to_string())
+                    .collect::<String>()
+            ),
+        };
+        writeln!(
+            f,
+            "DEFGATE {}{} AS {}:",
```

Comment:
Sounds good to me. My only counter-argument would be that `quil-rs` is hard coding what the default behavior should be instead of faithfully mirroring the specification. Ie. should `quil-rs` really be defining the default? Or should it omit and let a compiler or whatever else handle that? It's so unlikely to be relevant that it's not worth spending much time on and I think it's fine for `quil-rs` to have some opinions on the code it parses anyways.